text/0000-embroider-v2-package-format.md

```diff
@@ -0,0 +1,812 @@
+
+Allow you to test if an `import` (or `import()` or `importSync()`, since they all accept an argument with identical semantics) would succeed. Always compiles to a boolean literal.
+
+```js
+import { moduleExists, macroCondition, importSync } from "@ember/macros";
+if (macroCondition(moduleExists("ember-data"))) {
+  const DS = importSync("ember-data").default;
+  DS.Adapter.extend({
+    //
+  });
+}
+```
+
+Remember that you're always only allowed to `import` from your own **allowed dependencies**. So if an addon wants to optionally use another package only if that package is present, that package must be listed as an **Optional Peer Dependency**.
+
+`moduleExists` differs in one respect from all the other macros: there is no reliable build-time polyfill for it in non-Embroider builds. So when you use it in a non-Embroider build, you get a runtime implementation instead (equivalent to today's `require.has()`. If you need build-time branch elimination, try to use `dependencySatisfies` instead, which works correctly in both Embroider and non-Embroider builds.
+
+### JavaScript Macro: dependencySatisfies
+
+Allows you to test if the given **allowed dependency** satisfies the given semver range. Always compiles to a boolean literal.
+
+```js
+import { dependencySatisfies, macroCondition } from "@ember/macros";
+if (macroCondition(dependencySatisfies("ember-data", "^3.0.0"))) {
+  // include code here for ember-data 3.0 compat
+}
+```
+
+The package version will be `semver.coerce`d first, such that non-standard versions like "3.9.0-beta.0" will appropriately satisfy constraints like "> 3.8".
+
+### Javascript Macro: failBuild
+
+Allow you to cause a build failure with a custom error message. If `failBuild` isn't eliminated by `macroCondition`'s branch elimination, the build will fail.
+
+```js
+import { dependencySatisfies, macroCondition, failBuild } from "@ember/macros";
+if (macroCondition(dependencySatisfies("ember-data", "^3.0.0"))) {
+  // include code here for ember-data 3.0 compat
+} else {
+  failBuild(`We don't support ember-data versions other than ^3.0.0`);
+}
+```
+
+### Handlebars Macro: macroGetOwnConfig
+
+`macroGetOwnConfig` is very similar to the `getOwnConfig` JS macro, but it works as a Handlebars helper. Given this `OwnConfig`:
+
+```json
+{
+  "items": [{ "score": 42 }]
+}
+```
+
+Then:
+
+```hbs
+<SomeComponent @score={{macroGetOwnConfig "items" 0 "score" }} />
+{{! ⬆️compiles to ⬇️ }}
+<SomeComponent @score={{42}} />
+```
+
+If you don't pass any arguments, you can get the whole thing (although this makes your template bigger, so use arguments when you can):
+
+```hbs
+<SomeComponent @config={{macroGetOwnConfig}} />
+{{! ⬆️compiles to ⬇️ }}
+<SomeComponent @config={{hash items=(array (hash score=42))}} />
+```
+
+### Handlebars Macro: macroCondition
+
+Used as a helper within a block `{{#if}}` or inline `{{if}}`. Just like the JS `macroCondition`, it ensures that branch elimination will happen.
```

Comment:
This is still a little unclear to me. While working on a macro-based build system for Ember Bootstrap to handle Bootstrap version-specific builds I faced some open questions:

- Can another template helper be used as an argument for `macroCondition` as long as the value could be determined statically? E.g. `{{if (macroCondition (and (macroGetOwnConfig "isFoo") (macroGetOwnConfig "isBar")))}}`
- Are multiple positional arguments supported by `macroCondition`? E.g. `{{if (macroCondition (macroGetOwnConfig "isFoo") (macroGetOwnConfig "isBar"))}}`.
  - If it's not supported, a build time error should be thrown in my opinion.
  - If it's supported, are they combined by _AND_ or by _OR_?

Using `macroCondition` together with other template helpers would allow domain-specific macros without requiring an AST transform. For example in Ember Bootstrap we could use something like:

```hbs
{{#if (macroCondition (has-bootstrap-version 3))}}
  {{! Bootstrap 3 only }}
{{/if}}
```

With a `{{has-bootstrap-version}}` template helper like this:

```js
import { helper } from "@ember/component/helper";
import { getConfig } from "@ember/macros"; 

function hasBootstrapVersion([version]) {
  return getOwnConfig('version') === version;
}

export default helper(hasBootstrapVersion);
```

I don't see such a strong use case for support of multiple conditions. It would require all arguments to be statically determinable and in that case they could also be combined beforehand. So instead of `{{macroCondition (getOwnConfig "isFoo") (getOwnConfig "isBar)}}` a user could just add a `isFooAndBar` value to the config. If other template helpers are allowed as arguments even an `{{and}}` or `{{or}}` template helper could be used. But I nevertheless see the need to clarify that point.

Reply:
An arbitrary helper like `(has-bootstrap-version 3)` isn't going to work because we're not going to be able to run it at build time. But the intention is that you would instead put whatever code you want to run at build time in your `configure` build hook:

```js
configure() {
  let hasBootstrap3 = figureThisOutHowever();
  return { hasBootstrap3 };
}
```

And use the macro like:

```hbs
{{#if (macroCondition (macroGetOwnConfig "hasBootstrap3"))}}
  <UseSomeBootstrap3 />
{{/if}}
```

Reply:
As for boolean combinations, if Ember standardizes `and`, `or`, and `not` helpers I think it would be simple to support them here. But that is not anticipated in this RFC.

(The equivalent JS macro *does* support boolean operators.)